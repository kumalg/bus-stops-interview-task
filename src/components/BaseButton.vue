<template>
  <button :class="classes">
    <slot />
  </button>
</template>

<script setup lang="ts">
import { computed } from 'vue';

export type BUTTON_THEME = 'primary'

const props = withDefaults(defineProps<{
  theme?: BUTTON_THEME
  isActive?: boolean
}>(), {
  theme: 'primary',
  isActive: false
})

const classes = computed(() => {
  return ['base-button', `theme-${props.theme}`, { 'active': props.isActive }]
})
</script>

<style lang="scss" scoped>
.base-button {
  height: 2rem;
  padding: 0 1rem;
  display: inline-flex;
  align-items: center;
  font-size: 0.75rem;
  line-height: 1rem;
  user-select: none;
  border: none;
  border-radius: 0.25rem;
  font-family: inherit;
  
  @include default-transition('background-color, color');

  &:not(:disabled),
  &:not(.disabled) {
    cursor: pointer;
  }

  &.theme-primary {
    color: $color-gray-000;
    background-color: $color-accent;
  }

  &:hover, &.hover {
    &.theme-primary {
      background-color: $color-accent-2;
    }
  }

  &:active, &.active {
    &.theme-primary {
      background-color: $color-accent-3;
    }
  }
}
</style>