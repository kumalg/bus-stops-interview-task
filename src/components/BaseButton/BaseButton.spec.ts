import { flushPromises, shallowMount } from '@vue/test-utils';

import BaseButton from './BaseButton.vue';

const EXAMPLE_NAME = 'EXAMPLE_NAME';

describe('BaseButton', () => {
    const BaseButtonFactory = (
        props: InstanceType<typeof BaseButton>['$props'] = {},
        slotContent: string = EXAMPLE_NAME
    ) => {
        return shallowMount(BaseButton, {
            props,
            slots: {
                default: slotContent
            }
        });
    };

    it('renders proper html markup', () => {
        const wrapper = BaseButtonFactory();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('shows slot text', () => {
        const wrapper = BaseButtonFactory();

        expect(wrapper.text().trim()).toBe(EXAMPLE_NAME);
    });

    it('shows html markup inside', () => {
        const EXAMPLE_HTML = '<span class="test-span">TEST</span>';
        const wrapper = BaseButtonFactory({}, EXAMPLE_HTML);

        expect(wrapper.find('.test-span').exists()).toBe(true);
    });

    it('have theme primary class when no theme prop provided', () => {
        const wrapper = BaseButtonFactory();

        expect(wrapper.classes().includes('theme-primary')).toBe(true);
    });

    it('have theme primary class when theme prop is provided', () => {
        const wrapper = BaseButtonFactory({ theme: 'primary' });

        expect(wrapper.classes().includes('theme-primary')).toBe(true);
    });

    it('have active class when isActive is provided', async () => {
        const wrapper = BaseButtonFactory({ isActive: true });

        expect(wrapper.classes().includes('active')).toBe(true);

        wrapper.setProps({
            isActive: false
        });

        await flushPromises();

        expect(wrapper.classes().includes('active')).toBe(false);
    });
});
