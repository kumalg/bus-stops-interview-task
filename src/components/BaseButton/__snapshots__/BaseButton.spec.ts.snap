// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton renders proper html markup 1`] = `
<button
  class="base-button theme-primary"
>
  
  EXAMPLE_NAME
  
</button>
`;
